test/stop_on_request_test.cpp

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <unifex/stop_when.hpp>
+
+#include <unifex/defer.hpp>
+#include <unifex/let_done.hpp>
+#include <unifex/let_value_with_stop_source.hpp>
+#include <unifex/stop_on_request.hpp>
+#include <unifex/sync_wait.hpp>
+#include <unifex/then.hpp>
+#include <unifex/when_all.hpp>
+#include <unifex/when_all_range.hpp>
+
+#include <exception>
+#include <optional>
+
+#include <gtest/gtest.h>
+
+#if !UNIFEX_NO_COROUTINES
+
+#  include <unifex/single_thread_context.hpp>
+#  include <unifex/task.hpp>
+
+TEST(StopOnRequest, MultiThreadedCancellations) {
+  constexpr size_t iterations = 10;
+  constexpr size_t numSources = 5;
+  unifex::inplace_stop_source stopSources[iterations * numSources];
+
+  auto makeTask = [](unifex::inplace_stop_source& stopSource)
+      -> unifex::nothrow_task<void> {
+    stopSource.request_stop();
+    co_return;
+  };
+
+  bool wasCancelled = false;
+
+  for (size_t i = 0; i < iterations; i++) {
+    std::vector<unifex::any_sender_of<>> tasks;
+    unifex::single_thread_context threads[numSources];
+    for (size_t j = 0; j < numSources; j++) {
+      tasks.emplace_back(unifex::on(
+          threads[j].get_scheduler(),
+          makeTask(stopSources[i * numSources + j])));
+    }
+    auto cancellationSender = unifex::stop_on_request(
+                                  stopSources[i * numSources].get_token(),
+                                  stopSources[i * numSources + 1].get_token(),
+                                  stopSources[i * numSources + 2].get_token(),
+                                  stopSources[i * numSources + 3].get_token(),
+                                  stopSources[i * numSources + 4].get_token()) |
+        unifex::let_done([&]() {
+                                wasCancelled = true;
+                                return unifex::just();
+                              });
+    unifex::sync_wait(unifex::when_all(
+        unifex::when_all_range(std::move(tasks)), cancellationSender));
+    EXPECT_TRUE(wasCancelled);
+    wasCancelled = false;
+  }
```

Comment:
I think it's unfortunate that this test is being hidden from some CI jobs just because it depends on coroutines. Maybe we could rewrite it to work everywhere?
```suggestion
  constexpr size_t iterations = 10;
  constexpr size_t numSources = 5;

  for (size_t i = 0; i < iterations; i++) {
    std::array<unifex::single_thread_context, numSources> threads{};
    std::array<unifex::inplace_stop_source, numSources> stopSources{};

    auto makeTask = [&](size_t index) noexcept {
      auto& thread = threads[index];
      auto* stopSource = &stopSources[index];

      return unifex::on(thread, unifex::just_from([stopSource]() noexcept {
        stopSource->request_stop();
      });
    };

    std::vector<decltype(makeTask(0))> tasks;

    for (size_t j = 0; j < numSources; j++) {
      tasks.push_back(makeTask(j));
    }

    bool wasCancelled = false;
    auto cancellationSender = std::apply([&](auto&... stopSources) noexcept {
      return unifex::stop_on_request(stopSources.get_token()...)
          | unifex::let_done([&]() {
            wasCancelled = true;
            return unifex::just();
          });
    });

    unifex::sync_wait(unifex::when_all(
        unifex::when_all_range(std::move(tasks)), std::move(cancellationSender)));

    EXPECT_TRUE(wasCancelled);
  }
```